### PR TITLE
add support for authfilecredential

### PR DESCRIPTION
**Pull request by Luyunmt**
https://github.com/Azure/azure-sdk-for-net/pull/9312

 The purpose of this PR is explained in this or a referenced issue.
Add Azure AuthFile credential into azidentity.
 
 The PR does not update generated files. 
These files are managed by the codegen framework at [Azure/autorest.go][].
Don't include any generated files.
 
 The PR targets the latest branch.
This PR would target track2 branch.
 
 Tests are included and/or updated for code changes.
unit tests are included.
 
 Updates to [CHANGELOG.md][] are included.
No any packages changed, so it don't include the CHANGELOG.md
 
 Apache v2 license headers are included in each file.
Included
[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md

### CODE REVIEW

File: sdk/azidentity/auth_file_credential.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return &AuthFileCredential{filePath: filePath}, nil
+}
+
+// Obtains a token from the Azure Active Directory service, using the specified client detailed specified in the SDK Auth file.
+func (c *AuthFileCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	_, err := c.ensureCredential()
+	if err != nil {
+		return nil, &AuthenticationFailedError{msg: "Error parsing SDK Auth File: " + err.Error()}
```
Comment (ellismg) [inline sdk/azidentity/auth_file_credential.go]: NIT:  I think this should be `&AuthenticationFailedError{msg: "Error parsing SDK Auth File", inner: err}` so we don't drop the inner error (which may have additional context on top of what `.Error()` returns.

File: sdk/azidentity/auth_file_credential.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return c.credential, nil
+}
+
+// Parse credential file from local file path to byte.
+func (c *AuthFileCredential) parseCredentialsFile(filePath string) ([]byte, error) {
+	authData, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		log.Fatal(err)
```
Comment (ellismg) [inline sdk/azidentity/auth_file_credential.go]: I don't believe calling `log.Fatal(err)` is what we want here.  It will cause the process to exit  I think this should return `nil, err` if `error` is non `nil`.  If you agree, we could get rid of this method completely and just call `ioutil.ReadFile` in `ensureCredential`

File: sdk/azidentity/auth_file_credential.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Enables authentication to Azure Active Directory using configuration information stored Azure SDK Auth File.
+type AuthFileCredential struct {
+	filePath   string
+	credential azcore.TokenCredential
+}
+
+// Creates an instance of the SdkAuthFileCredential class based on information in given SDK Auth file.
+func NewAuthFileCredential(filePath string) (*AuthFileCredential, error) {
```
Comment (ellismg) [inline sdk/azidentity/auth_file_credential.go]: I think this also taken a `options *TokenCredentialOptions`, which are passed to the `NewClientSecretCredential` (but have the `AuthorityHost` that is used be the one specified in the file.  Similar to our other credential types.

File: sdk/azidentity/auth_file_credential.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	activeDirectoryEndpointUrl := token.ActiveDirectoryEndpointUrl
+
+	if clientId == "" || clientSecret == "" || tenantId == "" || activeDirectoryEndpointUrl == "" {
+		return nil, fmt.Errorf("Malformed Azure SDK Auth file. The file should contain 'clientId', 'clientSecret', 'tenantId' and 'activeDirectoryEndpointUrl' values.")
+	}
+
+	// Parse string activeDirectoryEndpointUrl to a Url.
+	AuthorityHost, err := url.Parse(activeDirectoryEndpointUrl)
+	if err != nil {
+		return nil, err
```
Comment (ellismg) [inline sdk/azidentity/auth_file_credential.go]: The addition of some context here might be helpful when trying to use the returned error to understand what went wrong.  Consider doing something like `return nil, fmt.Errorf("failed to parse activeDirectoryEndpointUrl in auth file: %w", err)`.

You could consider doing something similar on L95 to allow us to provide more context to the JSON decoding error.

File: sdk/azidentity/auth_file_credential.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	return c.credential.GetToken(ctx, opts)
+}
+
+// AuthenticationPolicy implements the azcore.Credential interface on AuthFileCredential.
+func (c *AuthFileCredential) AuthenticationPolicy(options azcore.AuthenticationPolicyOptions) azcore.Policy {
+	return newBearerTokenPolicy(c, options)
+}
+
+// Ensures that credential information is loaded from the SDK Auth file. This method should be called to initialize.
+func (c *AuthFileCredential) ensureCredential() (azcore.TokenCredential, error) {
```
Comment (ellismg) [inline sdk/azidentity/auth_file_credential.go]: I wonder if it's reasonable to pull all of the logic in `ensureCredential` into `NewAuthFileCredential`.  The fact it's a separate method was tied to how we did this in .NET (where doing blocking IO in a constructor felt a little weird) but it feels a little more idiomatic, but interested in what @jhendrixMSFT and other go folks think.

That would also let us front-load more error returning to when the credential is created instead of deferring until the fist time `GetToken` is called.

Comment (Luyunmt) [reply sdk/azidentity/auth_file_credential.go]: Since exception was thrown in getToken in other languages. We can do in same way here.

File: sdk/azidentity/auth_file_credential.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	TenantId                       string `json:"tenantId"`
+	ActiveDirectoryEndpointUrl     string `json:"activeDirectoryEndpointUrl"`
+	ResourceManagerEndpointUrl     string `json:"resourceManagerEndpointUrl"`
+	ActiveDirectoryGraphResourceId string `json:"activeDirectoryGraphResourceId"`
+	SqlManagementEndpointUrl       string `json:"sqlManagementEndpointUrl"`
+	GalleryEndpointUrl             string `json:"galleryEndpointUrl"`
+	ManagementEndpointUrl          string `json:"managementEndpointUrl"`
+}
+
+// Enables authentication to Azure Active Directory using configuration information stored Azure SDK Auth File.
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: The comment MUST start with AuthFileCredential to compile without errors.

File: sdk/azidentity/auth_file_credential.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+type AuthFileCredential struct {
+	filePath        string
+	credential      azcore.TokenCredential
+	authfileoptions *TokenCredentialOptions
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: Usually these options are stored by value so that they cannot be changed after the credential is created to avoid race conditions causing data corruption.

File: sdk/azidentity/auth_file_credential.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// Creates an instance of the SdkAuthFileCredential class based on information in given SDK Auth file.
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: Comment must start with "NewAuthFileCredential". Is there such a thing as SdkAuthFileCredential? I don't see it.

File: sdk/azidentity/auth_file_credential.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+func NewAuthFileCredential(filePath string, options *TokenCredentialOptions) (*AuthFileCredential, error) {
+	if filePath == "" {
+		return nil, fmt.Errorf("The parameter 'filePath' must set a value")
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: You mean it must not be the empty string; it has a value, it is "". I also think this check is not useful because someone can always write the code you have on line 42 and just pass "". So, if you have this check it should be where filePath is first used.

File: sdk/azidentity/auth_file_credential.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	return &AuthFileCredential{filePath: filePath, authfileoptions: options}, nil
+}
+
+// Obtains a token from the Azure Active Directory service, using the specified client detailed specified in the SDK Auth file.
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: Must start with GetToken

File: sdk/azidentity/auth_file_credential.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+func (c *AuthFileCredential) GetToken(ctx context.Context, opts azcore.TokenRequestOptions) (*azcore.AccessToken, error) {
+	_, err := c.ensureCredential()
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: Why does ensureCredential return a value that this ignores?
I see no other calls to this method that actually uses the 1st return value.

File: sdk/azidentity/auth_file_credential.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+	return c.credential.GetToken(ctx, opts)
+}
+
+// AuthenticationPolicy implements the azcore.Credential interface on AuthFileCredential.
+func (c *AuthFileCredential) AuthenticationPolicy(options azcore.AuthenticationPolicyOptions) azcore.Policy {
+	return newBearerTokenPolicy(c, options)
+}
+
+// Ensures that credential information is loaded from the SDK Auth file. This method should be called to initialize.
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: should be called by who to initialize?

File: sdk/azidentity/auth_file_credential.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return c.credential, nil
+}
+
+// Parse credential file from local file path to byte.
+func (c *AuthFileCredential) parseCredentialsFile(filePath string) ([]byte, error) {
+	authData, err := ioutil.ReadFile(filePath)
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: Why have a separate function with one line of code in it? You need the error check here AND also in the code that calls this duplicating the error checking.

File: sdk/azidentity/auth_file_credential.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	clientId := token.ClientId
+	clientSecret := token.ClientSecret
+	tenantId := token.TenantId
+	activeDirectoryEndpointUrl := token.ActiveDirectoryEndpointUrl
+
+	if clientId == "" || clientSecret == "" || tenantId == "" || activeDirectoryEndpointUrl == "" {
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: I'd kill these checks. There are all kinds of other checks we're not doing and these checks look for a specific kind of bug in the service but not all kinds of other potential service bugs. I think we can trust the service. If we can't, then we have much worse problems.

File: sdk/azidentity/auth_file_credential.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+		return nil, fmt.Errorf("Malformed Azure SDK Auth file. The file should contain 'clientId', 'clientSecret', 'tenantId' and 'activeDirectoryEndpointUrl' values.")
+	}
+
+	// Parse string activeDirectoryEndpointUrl to a Url.
+	AuthorityHost, err := url.Parse(activeDirectoryEndpointUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse activeDirectoryEndpointUrl in auth file: %w", err)
+	}
+	c.authfileoptions.AuthorityHost = AuthorityHost
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: It is scary to me to update option fields dynamically at runtime. If this is done, it MUST be done in a goroutine-safe way as other goroutines may be trying to use this client & its options at the same time.

File: sdk/azidentity/auth_file_credential.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+	return NewClientSecretCredential(tenantId, clientId, clientSecret, c.authfileoptions)
```
Comment (JeffreyRichter) [inline sdk/azidentity/auth_file_credential.go]: So, this is built by ensureCredential which it called by GetToken. This implies that each call to GetToken could create a new ClientSecretCredential which doesn't make sense. The logic flow is too hard to follow and opens the door for race conditions and creating garbage. I think the credential should be determined much earlier once, not in GetToken.

**Comment by Luyunmt** [comment]
Thanks for your review, We are looking into those comments carefully . As we were notified that @schabbs is finalizing the design. We will resolve all comments later.

File: sdk/azidentity/auth_file_credential.go (hunk 6 above)
Comment (Luyunmt) [reply sdk/azidentity/auth_file_credential.go]: Thanks for your review, We are looking into those comments carefully . As we were notified that @schabbs is finalizing the design. We will resolve all comments later.

**Comment by jongio** [comment]
@JeffreyRichter / @catalinaperalta - Can you please review these and help us get to sign-off for April release?  Thanks

**Comment by JeffreyRichter** [comment]
Done


  *   Jeffrey Richter (watch Architecting Distributed Cloud Apps<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fazure.microsoft.com%2Ftraining%2Fdistributed-cloud-apps%2F&data=02%7C01%7Cjeffreyr%40microsoft.com%7C82df4ff9e5ae46a5095908d7345e957b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637035453166233741&sdata=4GnkpFUKxlZXGXFr6oQo5GB%2FFNIbDXIZDXUij0q4AEE%3D&reserved=0>)

________________________________

File: sdk/azidentity/auth_file_credential.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+// AuthFileCredential enables authentication to Azure Active Directory using configuration information stored Azure SDK Auth File.
+type AuthFileCredential struct {
+	filePath        string
+	credential      azcore.TokenCredential
+	authfileoptions TokenCredentialOptions
+}
+
+// NewAuthFileCredential creates an instance of the AuthFileCredential class based on information in given SDK Auth file.
+func NewAuthFileCredential(filePath string, options *TokenCredentialOptions) (*AuthFileCredential, error) {
+	options, err := options.setDefaultValues()
```
[the remark this answers is not in the document]
Comment (Luyunmt) [reply sdk/azidentity/auth_file_credential.go]: The method setDefaultValues() I used is already exsited in azidentity.go . And other credential in 
method newAADIdentityClient in aad_identity_client.go use it.

[the remark this answers is not in the document]
Comment (JeffreyRichter) [reply sdk/azidentity/auth_file_credential.go]: @catalinaperalta & @jhendrixMSFT : Please look into setDefaultValue - it seems that this is not following our accepted options pattern.

**Comment by jongio** [comment]
This is on hold pending further discussion
